src/libs/antares/study/parts/thermal/container.cpp

```diff
@@ -118,7 +118,7 @@ uint PartThermal::prepareClustersInMustRunMode()
             if ((i->second)->mustrun)
             {
                 // Detaching the thermal cluster from the main list...
-                ThermalCluster* cluster = list.detach(i);
+                std::shared_ptr<ThermalCluster> cluster = list.detach(i);
```

Comment:
Here is what I missed : 
if you detach a (must-run) cluster from **list** to add it to **mustrunList**, then the cluster is no longer in **list**, but now in **mustrunList**.
So before this fix, destroying both lists should not cause a seg fault.
What did I miss ?

Reply:
Here is what happened previously :
1. A `shared_ptr` was created pointing to an object of type `ThermalCluster`, in container `list`.
2. For must-run clusters, this `shared_ptr` was detached from list and added to `musttunList`. For that, we use the `detach` method as above.
3. Formerly, this method returned a raw pointer to the `ThermalCluster` object, and called the destructor of `std::pair<String, std::shared_ptr<ThermalCluster>>`, which called the destructor for each member of the pair, which called the destructor for `ThermalCluster`.
4. `mustrunList` contained a `shared_ptr<ThermalCluster>` that pointed to freed memory. Since the destructor for `ThermalCluster` does nothing, that memory address still contained a valid object. Thus the simulation could take place. However, using deallocated memory is undefined behavior and may lead to segfaults.
5. The segfault actually happened at the destruction of `mustrunList`, after the simulation and writing of outputs.

The proposed solution is to avoid exposing raw pointers (step 3 above). By doing so, the destructor of `ThermalCluster` is called only once at the end of the simulation.

![image](https://user-images.githubusercontent.com/26088210/126952971-9ea399d4-54c4-4b5b-b8b3-8338bfd3a794.png)